wled00/wled.cpp

```diff
@@ -985,27 +996,58 @@ void WLED::handleConnection()
       sendImprovStateResponse(0x03, true);
       improvActive = 2;
     }
+    // try to reconnect if not able to connect for 30s or 18s
     if (now - lastReconnectAttempt > ((stac) ? 300000 : 18000) && wifiConfigured) {
       if (improvActive == 2) improvActive = 3;
       DEBUG_PRINTLN(F("Last reconnect too old."));
       if (++selectedWiFi >= multiWiFi.size()) selectedWiFi = 0; // we couldn't connect, try with another network from the list
       initConnection();
     }
+    // open AP if this is after boot or forced (_NO_CONN)
     if (!apActive && now - lastReconnectAttempt > 12000 && (!wasConnected || apBehavior == AP_BEHAVIOR_NO_CONN)) {
       if (!(apBehavior == AP_BEHAVIOR_TEMPORARY && now > WLED_AP_TIMEOUT)) {
         DEBUG_PRINTLN(F("Not connected AP."));
         initAP();  // start AP only within first 5min
       }
     }
-    if (apActive && apBehavior == AP_BEHAVIOR_TEMPORARY && now > WLED_AP_TIMEOUT && stac == 0) { // disconnect AP after 5min if no clients connected
+    // disconnect AP after 5min if no clients connected and mode TEMPORARY
+    if (apActive && apBehavior == AP_BEHAVIOR_TEMPORARY && now > WLED_AP_TIMEOUT && stac == 0) {
       // if AP was enabled more than 10min after boot or if client was connected more than 10min after boot do not disconnect AP mode
       if (now < 2*WLED_AP_TIMEOUT) {
+#ifndef WLED_DISABLE_ESPNOW
+        // we need to stop ESP-NOW as we are stopping AP
+        if (statusESPNow == ESP_NOW_STATE_ON) {
+          DEBUG_PRINTLN(F("ESP-NOW stopping on AP stop."));
+          quickEspNow.stop();
+          statusESPNow = ESP_NOW_STATE_UNINIT;
+          scanESPNow = now;
+        }
+#endif
         dnsServer.stop();
         WiFi.softAPdisconnect(true);
         apActive = false;
         DEBUG_PRINTLN(F("Temporary AP disabled."));
       }
     }
+#ifndef WLED_DISABLE_ESPNOW
+    // we are still not connected to WiFi and we may have AP active or not. if AP is active we will listen on its channel for ESP-NOW packets
+    // otherwise we'll try to find our master by hopping channels and wait 10s on each channel
+    if (!apActive && apBehavior == AP_BEHAVIOR_TEMPORARY && enableESPNow && wifiConfigured && !sendNotificationsRT) {
+      if (statusESPNow == ESP_NOW_STATE_UNINIT) {
+        quickEspNow.onDataSent(espNowSentCB);     // see udp.cpp
+        quickEspNow.onDataRcvd(espNowReceiveCB);  // see udp.cpp
+        DEBUG_PRINTLN(F("ESP-NOW initing in unconnected STA mode."));
+        bool espNowOK = quickEspNow.begin(channelESPNow); // use changeable channel STA mode
+        statusESPNow = espNowOK ? ESP_NOW_STATE_ON : ESP_NOW_STATE_ERROR;
+        scanESPNow = now;
+      }
+      if (statusESPNow == ESP_NOW_STATE_ON && now - scanESPNow > 10000) {
+        if (++channelESPNow > 13) channelESPNow = 1;
```

Comment:
thinking ahead...
today you store: 
* time of last scan
* current scanning channel

if instead you store:
* time of last beacon ping
* last beacon ping channel
you could make this a function `f(now - time of last ping) -> channel to try`

To be equivalent to a simple increment that would just be `(last ping channel + (t / 10000)) % 14 + 1`
Then in this function, you always recalculate the channel to try, and if it's != ESPNow's current channel, you switch to that channel.

This would introduce the ability to change the seek pattern easily in one function, for instance in the case where we are seeking but also want to listen to the last beacon channel more often, you could split time between an incrementing seek channel and the master channel.